firedrake/meshadapt.py

```diff
@@ -0,0 +1,113 @@
+from __future__ import absolute_import, print_function, division
+from six import with_metaclass
+import abc
+
+import numpy as np
+
+from firedrake.mesh import Mesh
+import firedrake.dmplex as dmplex
+import firedrake.utils as utils
+import firedrake.functionspace as functionspace
+import firedrake.function as function
+import firedrake.mesh as fmesh
+
+
+__all__ = ['adapt']
+
+
+class BaseAdaptation(with_metaclass(abc.ABCMeta)):
+    """
+    Abstract top-level class for mesh adaptation
+    """
+
+    def __init__(self, mesh):
+        """
+        """
+        self.mesh = mesh
+
+    @abc.abstractproperty
+    def newmesh(self):
+        pass
+
+    @abc.abstractmethod
+    def transfer_solution(self, f, **kwargs):
```

Comment:
I think we want to take a `*fields` argument, so I can pass any number of fields in.  The `kwargs` are method-specific parameters, which may be different depending on the concrete classes, so we just specify here that we can take arbitrary parameters, and the concrete classes name the ones they actually support.